--===========================================================================================================================>
--!strict
--===========================================================================================================================>

-- Set Memory Category:
debug.setmemorycategory('[Zoner]: Constants')

-- Return a Frozen Module Table:
return table.freeze({
	--====================================================================================================>
	
	Logs = table.freeze(require(script.Logs));
	
	TrackerData = {
		--=========================================================================================>
		
		-- Ignore due to their insignificance
		CharacterIgnoreParts = {
			--================================>
			UpperTorso = true,
			LowerTorso = true,
			Torso = true,
			LeftHand = true,
			RightHand = true,
			LeftFoot = true,
			RightFoot = true,
			--================================>
		}
		
		--=========================================================================================>
	};
	
	--====================================================================================================>
})

--===========================================================================================================================>